edk2basetools/VfrCompiler/IfrUtility.py

```diff
@@ -0,0 +1,2238 @@
+        except:
+            if pInFile is not None:
+                pInFile.close()
+            return None
+
+        PkgHeader = EFI_HII_STRING_PACKAGE_HDR.from_buffer(StringPtr)
+
+        # Check the String package
+        if PkgHeader.Header.Type != EFI_HII_PACKAGE_STRINGS:
+            return None
+
+        # Search the language, get best language based on RFC 4647 matching algorithm
+        Current = StringPtr
+        while not self.GetBestLanguage("en", PkgHeader.Language):
+            Current += PkgHeader.Header.Length
+            PkgHeader = EFI_HII_STRING_PACKAGE_HDR.from_buffer(Current)
+
+            # If can't find string package based on language, just return the first string package
+            if Current - StringPtr >= Length:
+                Current = StringPtr
+                PkgHeader = EFI_HII_STRING_PACKAGE_HDR.from_buffer(StringPtr)
+                break
+
+        Current += PkgHeader.HdrSize
+
+        # Find the string block according to the stringId
+        Status, NameOffset, BlockType = self.FindStringBlock(Current, StringId)
+        if Status != RETURN_SUCCESS:
+            return None
+
+        # Get varstore name according to the string type
+        if BlockType in [
+            EFI_HII_SIBT_STRING_SCSU,
+            EFI_HII_SIBT_STRING_SCSU_FONT,
+            EFI_HII_SIBT_STRINGS_SCSU,
+            EFI_HII_SIBT_STRINGS_SCSU_FONT,
+        ]:
+            StringName = Current + NameOffset
+            VarStoreName = StringName.decode("utf-8")
+        elif BlockType in [
+            EFI_HII_SIBT_STRING_UCS2,
+            EFI_HII_SIBT_STRING_UCS2_FONT,
+            EFI_HII_SIBT_STRINGS_UCS2,
+            EFI_HII_SIBT_STRINGS_UCS2_FONT,
+        ]:
+            UnicodeString = Current + NameOffset
+            Length = self.GetUnicodeStringTextSize(UnicodeString)
+            DestTmp = bytearray(Length // 2 + 1)
+            VarStoreName = DestTmp
+            index = 0
+            while UnicodeString[index] != "\0":
+                DestTmp[index] = ord(UnicodeString[index])
+                index += 1
+            DestTmp[index] = "\0"
+        else:
+            return None
+
+        return VarStoreName
+
+
+gVfrStringDB = VfrStringDB()
+
+EFI_RULE_ID_START = 0x01
+EFI_RULE_ID_INVALID = 0x00
+
+
+class SVfrRuleNode:
+    def __init__(self, RuleName=None, RuleId=0):
+        self.RuleId = RuleId
+        self.RuleName = RuleName
+        self.Next = None
+
+
+class VfrRulesDB(object):
+    def __init__(self):
+        self.RuleList = None
+        self.FreeRuleId = EFI_VARSTORE_ID_START
+
+    def RegisterRule(self, RuleName):
+        if RuleName == None:
+            return
+
+        pNew = SVfrRuleNode(RuleName, self.FreeRuleId)
+        if pNew == None:
+            return
+        self.FreeRuleId += 1
+        pNew.Next = self.RuleList
+        self.RuleList = pNew
+
+    def GetRuleId(self, RuleName):
+        if RuleName == None:
+            return
```

Comment:
You should not implicitly return None. You should either `return None` or return something, since you can possibly return EFI_RULE_ID_INVALID